core/store/src/sync_utils.rs

```diff
@@ -0,0 +1,72 @@
+use std::ops::{Deref, DerefMut};
+use std::sync::{Condvar, Mutex, MutexGuard};
+
+const POISONED_LOCK_ERR: &str = "The lock was poisoned.";
+
+/// A convenience wrapper around a SharedMutex and a Condvar.
+///
+/// It enables blocking while waiting for the underlying value to be updated.
+/// The implementation ensures that any modification results in all blocked
+/// threads being notified.
+pub(crate) struct Monitor<T> {
```

Comment:
Not saying that rolling out our own implementation is problematic but curious why we didn't use something off the shelf e.g. https://github.com/reem/rust-shared-mutex.

Reply:
I've considered considered using `shared-mutex` and actually my `Monitor` implementation is inspired by it.
Unfortunately it doesn't provide the API we need here, in particular we want `SharedMutexWriteGuard` to notify condvar when dropped (see [this comment](https://github.com/near/nearcore/pull/8215#pullrequestreview-1220506947)).

Reply:
Got it, thanks for the explanation.